### PR TITLE
Update Frontend Decouple

### LINKED ISSUE #764: Chalk'it .xprjson file life-cycle
For ensuring taipy-chalkit use-case, consider the simplest program:

```python
from taipy.gui import Gui
from taipy.chalkit import Page
 
a = 8
b = 10
c = a + b * 2
 
def on_change(state):
    state.c = state.a + state.b * 2

gui = Gui()
page = Page("my_project.xprjson")
gui.add_page("page", page)
gui.run(run_browser=True)
```

This program should:
- create an `my_project.xprjson` file at the same level as the calling module, with empty xprjson content, if this file does not exist
- if this file exists, or after its creation, its content should be provided to Chalk'it. A potential way is by extending the existing `taipyApp` with a new JavaScript function callback that could be named `loadProjectCallback`:

```javascript
const init = () => {
    // Init taipy app
    window.taipyApp = TaipyGuiBase.createApp(onInit);
    window.taipyApp.onChange = onChange;
    window.taipyApp.loadProjectCallback= loadProject;
```

An example of empty xprjson is the following:

```json
{
	"meta": {
		"version": "2.940.8734",
		"date": "Thu Jan 25 2024 23:26:07 GMT+0100 (heure normale d’Europe centrale)",
		"name": "my_project",
		"description": "",
		"groupName": "",
		"tags": [],
		"schedulerLogOff": true
	},
	"data": {
		"datanodes": []
	},
	"libraries": {
		"pyodideStandard": [],
		"pyodideMicropip": []
	},
	"scaling": {
		"widthPx": 1295,
		"heightPx": 590,
		"scrollWidthPx": 1295,
		"scrollHeightPx": 590,
		"widthVw": 84.75130890052355,
		"heightVh": 78.66666666666667,
		"scrollWidthVw": 84.75130890052355,
		"scrollHeightVh": 78.66666666666667,
		"scalingMethod": "scaleTwh",
		"media": "large",
		"colDims": {
			"widthPx": 1295.4,
			"heightPx": 590.4,
			"widthVw": 84.77748691099477,
			"heightVh": 78.72,
			"rowHeightPercent": 100
		}
	},
	"device": {
		"cols": {
			"valueRow": "1",
			"valueCol": "1",
			"maxCells": 1,
			"maxCols": 1,
			"classType": "col-md-12"
		},
		"droppers": {
			"dpr1": {}
		},
		"backgroundColor": "",
		"theme": "default"
	},
	"dashboard": {},
	"connections": {},
	"exportOptions": "projectToTargetWindow",
	"pages": {
		"pageNames": [
			"Page 1"
		],
		"defaultPage": {}
	},
	"checkExportOptions": true,
	"navBarNotification": false
}
```

A JavaScript `saveProject` function needs also to be provided

```JavaScript
window.taipyApp.saveProject(project_name, xprjson_content);
```

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Resolves #764 #766

### CODE REVIEW

**Review by FredLL-Avaiga — DISMISSED** [review]

File: frontend/taipy-gui/base/src/app.ts [hunk 1]
```diff
@@ -39,19 +40,19 @@ export class TaipyApp {
     }
     set onInit(handler: OnInitHandler | undefined) {
         if (handler !== undefined && handler?.length !== 1) {
-            throw new Error("onInit function requires 1 parameter")
+            throw new Error("onInit function requires 1 parameter");
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/app.ts]:
```suggestion
            throw new Error("onInit() requires one parameter");
```

File: frontend/taipy-gui/base/src/app.ts [hunk 2]
```diff
@@ -39,19 +40,19 @@ export class TaipyApp {
...
+        this._onInit = handler;
     }
 
     get onChange() {
         return this._onChange;
     }
     set onChange(handler: OnChangeHandler | undefined) {
         if (handler !== undefined && handler?.length !== 3) {
-            throw new Error("onChange function requires 3 parameters")
+            throw new Error("onChange function requires 3 parameters");
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/app.ts]:
```suggestion
            throw new Error("onChange() requires three parameters");
```

File: frontend/taipy-gui/base/src/app.ts [hunk 3]
```diff
@@ -95,6 +96,15 @@ export class TaipyApp {
         }
         sendWsMessage(this.socket, "GMC", "get_module_context", { path: path }, this.clientId);
     }
+
+    trigger(actionName: string, triggerId: string, payload: Record<string, unknown> = {}) {
+        payload["action"] = actionName;
+        sendWsMessage(this.socket, "A", triggerId, payload, this.clientId, this.context);
+    }
+
+    uploadFile(encodedName: string, files: FileList, progressCallback: (val: number) => void) {
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/app.ts]: Rename to 'upload()'? I don't see the need for 'File'

File: frontend/taipy-gui/base/src/app.ts [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import { sendWsMessage } from "../../src/context/wsUtils";
+import { uploadFile } from "../../src/workers/fileupload";
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/app.ts:2]: Rename file to 'upload'?

File: frontend/taipy-gui/base/src/app.ts [hunk 5]
```diff
@@ -95,6 +96,15 @@ export class TaipyApp {
...
         sendWsMessage(this.socket, "GMC", "get_module_context", { path: path }, this.clientId);
     }
+
+    trigger(actionName: string, triggerId: string, payload: Record<string, unknown> = {}) {
+        payload["action"] = actionName;
+        sendWsMessage(this.socket, "A", triggerId, payload, this.clientId, this.context);
+    }
+
+    uploadFile(encodedName: string, files: FileList, progressCallback: (val: number) => void) {
+        return uploadFile(encodedName, files, progressCallback, this.clientId);
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/app.ts:106]: Same, maybe?

**Review by FabienLelaquais — DISMISSED** [review]
👌

File: frontend/taipy-gui/base/src/app.ts (hunk 5 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/app.ts:106]: I will keep what is already in the product

File: frontend/taipy-gui/base/src/app.ts (hunk 4 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/app.ts:2]: I will keep this one too

File: frontend/taipy-gui/base/src/app.ts (hunk 3 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/app.ts]: I will rename for this one! tx Fabien

**Review by FabienLelaquais — APPROVED** [review]
👍

**Review by FredLL-Avaiga — APPROVED** [review]

**Pull request merged**